### ISSUE TITLE
onRun hook is not fired EVER AGAIN when a hot code push happens

### ISSUE THREAD

**Issue by Sewdn** [iron-meteor/iron-router#1219]
The onRun hook is supposed to be fired when the route is ran, but not when the route is reran because of a computation invalidation. The doc states clearly that the onRun callback will not be ran when a hot code push reloads the page.

However, as soon as your app reloads because of a hotcode push, the onRun hook is never called again! Not even when you navigate to new routes. As soon as a route is activated, you expect this route to run the onRun callback (once), whether a hot code push has happened or not during your app runtime, no?

This can be a problem for a lot of people tracking their analytics on an onRun hook (a suggested in the docs). As soon as you deploy, all connected users will not trigger any analytics until they refresh.

When you add crucial functionality in the onRun callback, your app can break after deployment for connected users.

Also, during development, this issue is triggered after every file you save.

I added a very simple app with only iron-router as non-core package to be able to reproduce the issue:
https://github.com/Sewdn/iron-router-onrun

Thanks for looking into this!
Kr

**Comment by funkyeah** [comment]
I can confirm I am also seeing this

**Comment by dcsan** [comment]
ok, that's what it is? I thought it was my mistake. guess this type of thing is hard to write tests for...

http://stackoverflow.com/questions/28728705/ironrouter-onrun-doesnt-fire

```
Router.map ->

  @route 'slots/tags',
    path: "/slots/:tags"
    data: ->
      console.log("IR:data")
      blob = {
        params: @params
        tags: @params.tags
      }
      window.blob = blob
      return blob

    # wtf it doesnt seem to run?
    onRun: ->
      console.log("IR:onRun")

    action: ->
      console.log("IR:action")
      createSlotsGame(@)
      @render 'slots'
```

**Comment by pauloborges** [comment]
+1

Is there a temporary solution for this issue?

**Comment by pauloborges** [comment]
This is probably related to #1130.

**Comment by cmather** [comment]
This is by design. onRun is intended to run once, when the route first
runs. Use the action function or the onBeforeAction hook if you want
indefinite reactivity.

**Comment by pauloborges** [comment]
Please check this reproduction: https://github.com/pauloborges/iron-router-onrun-bug

You'll be able to see that on both routes, `/test` and `/test2`, the `onRun` method is being called. Now do this test: open `/test` and do a hot code push by, for example, changing `Test`'s template. If you hit back and navigate to `/test2` by clicking the second button, its `onRun` method won't be called.

Is this expected? I guess it's not. As @Sewdn said, if one hot code push happened, it will break all route's `onRun` methods.

**Comment by funkyeah** [comment]
@cmather: What Paul, I, and others are seeing is not that `onRun` doesn't run immediately after a hot code reload, but rather that onRun no longer executes for `any` future routes after a hot code reload.

e.g.
 1. navigate to route 1 -> onRun runs
 2. navigate away and then back to route 1 -> onRun runs
 3. hotcode reload -> onRun doesn't run (this is fine)
 4. navigate away and then back to route 1 -> (onRun doesn't run (this is the problem))

**Comment by cmather** [comment]
Oh I see. That sounds buggish. Have you guys identified where the hot code stuff happens to take a crack at it?

**Comment by dcsan** [comment]
i realized in my own code i have an onRun in the main controller. so this could be overriding/overwriting anything defined at a route level. i assumed it would be adding to the list of onRuns, kind of like how Meteor.onStartup works.

**Comment by pauloborges** [comment]
Apparently `_hasJustReloaded` never gets a falsey value after the first hot code push, so this condition is never true:

```
function onRun (req, res, next) {
  if (this._computation.firstRun && !RouteController._hasJustReloaded) { //here
    RouteController._hasJustReloaded = false;
    if (onRunStack.length > 0) {
      onRunStack.dispatch(req.url, this, next);
    } else {
      next();
    }
  } else {
    next();
  }
}
```

Source: https://github.com/iron-meteor/iron-router/blob/17b30849f50d618f99883adc47ec9b63adfc195b/lib/route_controller_client.js#L197-L208

The other place where the variable is written is below, but I don't know how `Meteor._reload` works.
```
if (Meteor._reload) {
  // just register the fact that a migration _has_ happened
  Meteor._reload.onMigrate('iron-router', function() { return [true, true]});

  // then when we come back up, check if it it's set
  var data = Meteor._reload.migrationData('iron-router');
  RouteController._hasJustReloaded = data;
}
```

Source: https://github.com/iron-meteor/iron-router/blob/17b30849f50d618f99883adc47ec9b63adfc195b/lib/route_controller_client.js#L370-L377

**Comment by cmather** [comment]
@pauloborges, yep that's the place. And it looks like there might be some issue between _hasJustReloaded being set on the RouteController vs. the instance. But I'd have to spend time reasoning through it.

For @pauloborges or anyone else who wants to take a stab at fixing this: If you want to take 28 minutes to learn about hot code reload you can watch this class: https://www.eventedmind.com/classes/hot-code-push-with-reload-and-autoupdate. Email me and I'll give you a coupon where you can watch for free :).

**Comment by fabienb4** [comment]
@cmather I was having problems due to that bug with `onRun` not being called after hot code pushes. I tried a small fix, it works for me... I created a [test repo](https://github.com/fabienb4/iron-router-onrun-test) if you want to try it.
The value from `RouteController._hasJustReloaded` is never used anywhere else but for the `onRun` function, so it shouldn't cause any other problem.

The change is that simple: 
```javascript
    function onRun (req, res, next) {
      if (this._computation.firstRun && !RouteController._hasJustReloaded) {
        RouteController._hasJustReloaded = false;
        if (onRunStack.length > 0) {
          onRunStack.dispatch(req.url, this, next);
        } else {
          next();
        }
      } else {
        next();
      }
    },
```

to: 
```javascript
    function onRun (req, res, next) {
      if (this._computation.firstRun && !RouteController._hasJustReloaded) {
        if (onRunStack.length > 0) {
          onRunStack.dispatch(req.url, this, next);
        } else {
          next();
        }
      } else {
        next();
      }
      RouteController._hasJustReloaded = false;
    },
```
Oh, and in the test repo, I also updated the Meteor._reload code, which is now deprecated ([Meteor reload package](https://github.com/meteor/meteor/blob/devel/packages/reload/deprecated.js)): 
```javascript
if (Package.reload) {
  // just register the fact that a migration _has_ happened
  Package.reload.Reload._onMigrate('iron-router', function() { return [true, true] });

  // then when we come back up, check if it is set
  var data = Package.reload.Reload._migrationData('iron-router');
  RouteController._hasJustReloaded = data;
}
```

**Comment by joegoldbeck** [comment]
+1

**Comment by DenisGorbachev** [comment]
@cmather Could you please merge the pull request by @fabienb4? It's clearly a bug, and the proposed fix looks pretty straightforward.

**Comment by cmather** [comment]
Yeah I'll add some time this week.

**Comment by DenisGorbachev** [comment]
Thanks a lot!

**Comment by steph643** [comment]
+1

**Comment by nakhbari** [comment]
+1 Please merge the pull request by @fabienb4!

**Comment by dcsan** [comment]
could anyone else be added to the IR team that has more time for these kind of things? 

I guess with the state of meteor testing it makes it really hard to merge fixes for OSS projects, and if MDG don't want to maintain a router in core, we're down to 4 month wait times for PRs to be merged into IR.

**Comment by dstollie** [comment]
+1 I don't like the idea of using a router which isn't being maintained this bad

**Comment by atulmy** [comment]
:/ waiting!

**Comment by liangsun** [comment]
Is this bug resolved? How's it going? I got the same issue.

**Comment by fabienb4** [comment]
270 issues, last commit a month ago (update readme)... I guess the author is busy, but it would be nice if someone could pick it up...
I'm not using it in my current projects, but people may need it in theirs.

PS: Look at the combo Session variables, dynamic template inclusion & template subscriptions. It might suit your needs, it works even better for mine...

**Comment by teddis** [comment]
I don't know if this bug has been resolved yet. (I cannot update from iron:router 1.0.9 to 1.0.12 due to the dreaded "isobuild" bug which also appears not to have been fixed.) Anyway, I solved this for my case with a workaround using redirection. Request /a, call onBeforeAction(), then redirect to /b. For me, this one particular request then can be logged but not accidentally relogged if the user refreshes the browser or a hot code push occurs. This doesn't really scale well for more routes. I wish these semi-critical bugs could be fixed by MDG and more rapidly if they are to gain credibility and developer buy-in of the platform. I know iron:router package is not officially part of Meteor, but seems like it ought to be given its importance and high use penetration.
